core/drivers/clk/clk.c

```diff
@@ -310,3 +321,108 @@ TEE_Result clk_get_rates_array(struct clk *clk, size_t start_index,
 
 	return clk->ops->get_rates_array(clk, start_index, rates, nb_elts);
 }
+
+/* Return updated message buffer position of NULL on failure */
+static __printf(3, 4) char *add_msg(char *cur, char *end, const char *fmt, ...)
+{
+	va_list ap = { };
+	int max_len = end - cur;
+	int ret = 0;
+
+	va_start(ap, fmt);
+	ret = vsnprintf(cur, max_len, fmt, ap);
+	va_end(ap);
+
+	if (ret < 0 || ret >= max_len)
+		return NULL;
+
+	return cur + ret;
+}
+
+static void __maybe_unused print_clock(struct clk *clk, int indent)
+{
+	static const char * const rate_unit[] = { "Hz", "kHz", "MHz", "GHz" };
+	int rate_unit_index = 0;
+	unsigned long rate = 0;
+	char msg_buf[128] = { };
+	char *msg_end = msg_buf + sizeof(msg_buf);
+	char *msg = msg_buf;
+	int i = 0;
+
+	/*
+	 * Current prints the clock state based on the clock refcount.
+	 * A futur change could print the hardware clock state when
+	 * related clock driver provides a struct clk_ops::is_enabled handler
+	 */
+
+	if (indent) {
+		for (i = 0; i < indent - 1; i++) {
+			msg = add_msg(msg, msg_end, "|   ");
+			if (!msg)
+				goto out;
+		}
+
+		msg = add_msg(msg, msg_end, "+-- ");
+		if (!msg)
+			goto out;
+	}
+
+	rate = clk_get_rate(clk);
+	if (rate == (rate / 1000 * 1000)) {
```

Comment:
Isn't this the same as `if (rate % 1000 == 0)`?
Also, I'd prefer to have better symmetry with the ifs like:
``` C
if (rate % 1000) {
         rate_unit_index = 0;
} else if (rate % 10000000) {
         rate_unit_index = 1;
         rate /= 1000000;

} else ...
```
You could even turn that into a loop.

Reply:
Thanks for `%` suggestion. Cleaner indeed.

Regarding using a loop, I wonder if using a loop will make this 3 cases if/else more readable. I'll give a try.